temporalcli/commands_test.go

```diff
@@ -0,0 +1,394 @@
+package temporalcli_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"slices"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"github.com/temporalio/cli/temporalcli"
+	"github.com/temporalio/cli/temporalcli/devserver"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/worker"
+	"go.temporal.io/sdk/workflow"
+)
+
+type CommandHarness struct {
+	*require.Assertions
+	t       *testing.T
+	Options temporalcli.CommandOptions
+	// Defaults to a context closed on close or test complete
+	Context context.Context
+	// Can be used to cancel context given to commands (simulating interrupt)
+	CancelContext context.CancelFunc
+}
+
+func NewCommandHarness(t *testing.T) *CommandHarness {
+	h := &CommandHarness{Assertions: require.New(t), t: t}
+	h.Context, h.CancelContext = context.WithCancel(context.Background())
+	t.Cleanup(h.Close)
+	return h
+}
+
+// Reentrant, called after test by default, cancels context
+func (h *CommandHarness) Close() {
+	// Cancel context
+	if h.CancelContext != nil {
+		h.CancelContext()
+	}
+}
+
+func (h *CommandHarness) ContainsOnSameLine(text string, pieces ...string) {
+	// Split into lines, then check each piece is present
+	lines := strings.Split(text, "\n")
+	for _, line := range lines {
+		foundAll := true
+		for _, piece := range pieces {
+			if !strings.Contains(line, piece) {
+				foundAll = false
+				break
+			}
+		}
+		if foundAll {
+			return
+		}
+	}
+	h.Fail("Pieces not found on any line together")
```

Comment:
This will return `true` with the pieces in any order, but I don't think that's what we want. I think we should consider using a `ContainsLineMatchingRegex` utility for many of the assertions.

Reply:
I am ok with any order as I'm trying for very loose assertion here of stdout (otherwise it becomes brittle). But I expect as commands and tests are developed, more and more assertion helpers are written.

Reply:
> I am ok with any order as I'm trying for very loose assertion here of stdout

Can you give an example of when it would be appropriate to not care about the order of the substrings in the output? If there aren't examples of that then please change to requiring order. It's very unusual to make substring assertions but not care about the order of the substrings.

Reply:
> When I wrote these tests I didn't care about order, so the examples are there

Right, but I asked for examples of when it was appropriate to ignore order, and I would say that in those examples it wasn't :) IMO all these clearly should be asserting order:

```
	s.ContainsOnSameLine(out, "1", "WorkflowExecutionStarted")
	s.ContainsOnSameLine(out, "2", "WorkflowTaskScheduled")
	s.ContainsOnSameLine(out, "3", "WorkflowTaskStarted")
	s.ContainsOnSameLine(out, "Status", "COMPLETED")
	s.ContainsOnSameLine(out, "Result", `{"foo":"bar"}`)
```

> I also didn't care about spacing between them, what their indentation was, what the line above/below them is, etc. It can be said it's unusual to make assertions about data on a line without caring about the spacing or indentation too.

Disagree. Order of words has vastly more impact on semantics than spacing and indentation.